src/qibo/quantum_info/quantum_networks.py

```diff
@@ -431,15 +430,22 @@ def __mul__(self, number: Union[float, int]):
                 "It is not possible to multiply a ``QuantumNetwork`` by a non-scalar.",
             )
 
-        matrix = self._full()
-
-        return QuantumNetwork(
-            number * matrix,
-            self.partition,
-            self.system_output,
-            self._pure,
-            self._backend,
-        )
+        if self.pure():
+            return QuantumNetwork(
+                np.emath.sqrt(number) * self.matrix(),
```

Comment:
@Canoming do you need `np.emath.sqrt` here to return complex numbers or does `np.sqrt` suffice?